# TM10007 Assignment template

In [16]:
# Run this to use from colab environment
# %pip install -q --upgrade git+https://github.com/jveenland/tm10007_ml.git

Note: you may need to restart the kernel to use updated packages.


In [37]:
# %pip install -U scikit-learn

  Attempting uninstall: joblib
    Found existing installation: joblib 1.1.0
    Uninstalling joblib-1.1.0:
      Successfully uninstalled joblib-1.1.0
  Attempting uninstall: scikit-learn
    Found existing installation: scikit-learn 1.0.2
    Uninstalling scikit-learn-1.0.2:
      Successfully uninstalled scikit-learn-1.0.2
Note: you may need to restart the kernel to use updated packages.


## Data loading and cleaning

Below are functions to load the dataset of your choice. After that, it is all up to you to create and evaluate a classification method. Beware, there may be missing values in these datasets. Good luck!

In [13]:
# Data loading functions. Uncomment the one you want to use
from worcliver.load_data import load_data
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

data = load_data()

# Data clean-up (before splitting data)
In this section double entries, such as repeated features and samples, will be removed from the dataset.

In [4]:
df = pd.DataFrame(data)

'''
These 2 lines add a duplicate sample in the DataFrame, you can than observe what the effect is 
in the printed sentence. If you'd like to see it indeed works, please uncomment these lines.
'''
# row = df.iloc[0].copy()
# df = df.append(row, ignore_index=True)
'''
In the lines below we first drop all duplicate entries, whether it is a duplicate row or a column. 
Additionally, the number of duplicates will be returned to the user.
'''

df_f = df.drop_duplicates()
print(f'The number of samples (after removal of duplicates): {len(df_f.index)}, the number of duplicates that had to be removed: {len(df.index)-len(df_f.index)}')
print(f'The number of columns (after removal of duplicates): {len(df_f.columns)}, the number of duplicates that had to be removed: {len(df.columns)-len(df_f.columns)}')

The number of samples (after removal of duplicates): 186, the number of duplicates that had to be removed: 0
The number of columns (after removal of duplicates): 494, the number of duplicates that had to be removed: 0


In [57]:
df = df_f                                       # So we can just use df from now on
df = df.replace('', np.nan)                     # Any empty values are replaced for NaN
df.describe(include='all')                    # Uncomment if interested

label  PREDICT_original_sf_compactness_avg_2.5D  \
count         186                                186.000000   
unique          2                                       NaN   
top     malignant                                       NaN   
freq           94                                       NaN   
mean          NaN                                  0.808754   
std           NaN                                  0.070161   
min           NaN                                  0.549046   
25%           NaN                                  0.773570   
50%           NaN                                  0.824074   
75%           NaN                                  0.857698   
max           NaN                                  0.926346   

        PREDICT_original_sf_compactness_std_2.5D  \
count                                 186.000000   
unique                                       NaN   
top                                          NaN   
freq                                         NaN   
mean                                    0.071922   
std                                     0.047649   
min                                     0.004729   
25%                                     0.041594   
50%                                     0.056596   
75%                                     0.082090   
max                                     0.245769   

        PREDICT_original_sf_rad_dist_avg_2.5D  \
count                              186.000000   
unique                                    NaN   
top                                       NaN   
freq                                      NaN   
mean                                30.933690   
std                                 14.209702   
min                                  9.648695   
25%                                 19.942274   
50%                                 27.275579   
75%                                 38.754607   
max                                 73.610032   

        PREDICT_original_sf_rad_dist_std_2.5D  \
count                              186.000000   
unique                                    NaN   
top                                       NaN   
freq                                      NaN   
mean                                 4.662412   
std                                  2.656027   
min                                  1.113733   
25%                                  2.762641   
50%                                  3.986081   
75%                                  6.093413   
max                                 16.559442   

        PREDICT_original_sf_roughness_avg_2.5D  \
count                               186.000000   
unique                                     NaN   
top                                        NaN   
freq                                       NaN   
mean                                  8.502448   
std                                   3.301010   
min                                   2.310309   
25%                                   6.407817   
50%                                   7.850429   
75%                                  10.244457   
max                                  21.955618   

        PREDICT_original_sf_roughness_std_2.5D  \
count                               186.000000   
unique                                     NaN   
top                                        NaN   
freq                                       NaN   
mean                                  3.342782   
std                                   2.233298   
min                                   0.152609   
25%                                   1.925876   
50%                                   2.826361   
75%                                   4.118141   
max                                  15.637597   

        PREDICT_original_sf_convexity_avg_2.5D  \
count                               186.000000   
unique                                     NaN   
top                                        NaN   
freq                                       NaN   


### Conclusion
The table above shows there are only 2 labels present in the dataset, from observation of the csv.file we can see the labels are 'malignant' and 'benigne'. What is also stated here, that 94 of the 186 samples was labeled malignant (and thus 92 samples are labeled benigne.) This information is usefull in the next step, to determine whether a stratified splitting of our data, would be necessary.

# Data splitting
In this section the dataset will be split into two datasets, the Design set (D_set) and the Final Test set (Ft_set). This splitting will be performed randomly where 80% will end up in D_set and the remaining 20% will be stored in Ft_set. Ft_set will not be used at all, until the very latest to evaluate our tool and establish how generalising the tool is. In the initial split we will randomly split the dataset, as both labels are equally abundant in the set.

In [7]:
from sklearn.model_selection import train_test_split
# D_set, Ft_set = train_test_split(df, test_size=0.2)
# D_set = pd.DataFrame(D_set)
# Ft_set = pd.DataFrame(Ft_set)                         

In [21]:
# D_set.to_csv("C:/Users/dgjpa/D_set.csv")
# Ft_set = Ft_set.to_csv("C:/Users/dgjpa/Ft_set.csv")

'''
These lines were used to save the splitted sets, to make sure Design and Test set won't alter.
''' 

"\nThese lines were used to save the splitted sets, to make sure Design and Test set won't alter.\n"

# Data splitting (2/2)
First, load_sets.py there will be two functions who returns a dataframe from the save .csv file. D_usable and Ft_usable are both DataFrames including both the labels as the feature information, and are the resulting datasets from the initial split as can be seen above.

X_d is a Dataframe including all the information about a case but with the labels removed, whereas y_d is a DataFrame including only the labels.

### !!! Important v
Still unclear: Is whether it is necessary to manually split the X_d & y_d DataFrames already in X_train/y_train and X_test/y_test sets

This second split is ofcourse necessary to perform cross-validation, however several sklearn-functions seem to have this split already incorporated in their functions, such as: cross_validate, cross_val_score(Seems to be especially handy)

In [3]:
from function_files.load_sets import load_D_set
from sklearn.model_selection import train_test_split
D_s = load_D_set()
D_s.head()

label  ...  PREDICT_original_phasef_phasesym_entropy_WL3_N5
ID                      ...                                                 
Liver-162_0  malignant  ...                                        15.152812
Liver-107_0  malignant  ...                                        12.426019
Liver-027_0  malignant  ...                                        15.656969
Liver-012_0     benign  ...                                        11.009890
Liver-028_0  malignant  ...                                        11.361561

[5 rows x 494 columns]

In [56]:
D_s.describe(include='all')

label  PREDICT_original_sf_compactness_avg_2.5D  \
count      148                                148.000000   
unique       2                                       NaN   
top     benign                                       NaN   
freq        75                                       NaN   
mean       NaN                                  0.810449   
std        NaN                                  0.071607   
min        NaN                                  0.549046   
25%        NaN                                  0.774140   
50%        NaN                                  0.825621   
75%        NaN                                  0.860922   
max        NaN                                  0.926346   

        PREDICT_original_sf_compactness_std_2.5D  \
count                                 148.000000   
unique                                       NaN   
top                                          NaN   
freq                                         NaN   
mean                                    0.070357   
std                                     0.048925   
min                                     0.004729   
25%                                     0.040098   
50%                                     0.054622   
75%                                     0.080052   
max                                     0.245769   

        PREDICT_original_sf_rad_dist_avg_2.5D  \
count                              148.000000   
unique                                    NaN   
top                                       NaN   
freq                                      NaN   
mean                                31.025697   
std                                 14.232578   
min                                  9.648695   
25%                                 20.157877   
50%                                 27.523779   
75%                                 38.818462   
max                                 73.610032   

        PREDICT_original_sf_rad_dist_std_2.5D  \
count                              148.000000   
unique                                    NaN   
top                                       NaN   
freq                                      NaN   
mean                                 4.674168   
std                                  2.753151   
min                                  1.130516   
25%                                  2.747720   
50%                                  3.986081   
75%                                  6.072275   
max                                 16.559442   

        PREDICT_original_sf_roughness_avg_2.5D  \
count                               148.000000   
unique                                     NaN   
top                                        NaN   
freq                                       NaN   
mean                                  8.464892   
std                                   3.356987   
min                                   2.310309   
25%                                   6.380224   
50%                                   7.914242   
75%                                  10.175131   
max                                  21.955618   

        PREDICT_original_sf_roughness_std_2.5D  \
count                               148.000000   
unique                                     NaN   
top                                        NaN   
freq                                       NaN   
mean                                  3.333409   
std                                   2.270701   
min                                   0.265697   
25%                                   1.899252   
50%                                   2.838249   
75%                                   4.034999   
max                                  15.637597   

        PREDICT_original_sf_convexity_avg_2.5D  \
count                               148.000000   
unique                                     NaN   
top                                        NaN   
freq                                       NaN   
mean                             

In [58]:
D_s.info()

<class 'pandas.core.frame.DataFrame'>
Index: 148 entries, Liver-162_0 to Liver-008_0
Columns: 494 entries, label to PREDICT_original_phasef_phasesym_entropy_WL3_N5
dtypes: float64(468), int64(25), object(1)
memory usage: 572.3+ KB


Below, we have splitted the Design set into a dataframe including the label, and a dataframe includinng only the information without label.

In [4]:
from function_files.load_sets import load_D_set
from sklearn.model_selection import train_test_split
D_s = load_D_set()
y_d = D_s.pop('label').to_frame()
X_d = D_s

In [5]:
X_d.shape, y_d.shape

((148, 493), (148, 1))

## We will still need to split the X_d and y_d in a training and a validation set